src/cephadm/cephadm.py

```diff
@@ -2703,6 +2703,35 @@ def _rollback(ctx: CephadmContext) -> Any:
             raise
     return cast(FuncT, _rollback)
 
+def get_python_version(ctx: CephadmContext):
+    """
+    Retrieves the Python version from a specified daemon's container.
+    """
+    daemon_types = ['mon', 'mgr', 'osd']
+    setattr(ctx,'python_version',None)
+    for daemon in daemon_types:
+        container_info = get_container_info(ctx, daemon, by_name=False)
+        if container_info and container_info.container_id:
+            container_id = container_info.container_id
+            try:
+                exec_command = [
+                    "podman", "exec", container_id, "python3",
+                    "-c", "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')"
+                ]
+                exec_result = subprocess.run(exec_command, capture_output=True, text=True)
+                python_version = exec_result.stdout.strip()
+        
+                # Store the detected Python version in ctx
+                if not ctx.python_version:
+                    ctx.python_version = python_version
+                else:
+                    if ctx.python_version == python_version:
+                        logger.info("Retrieved python version matches the stored version; no update")
+                    else:
+                        logger.warning("Python version mismatch")
+                logger.info(f"Detected Python version {python_version} in container {container_id}")
+            except Exception as e:
+                logger.error(f"Failed to get Python version for container {container_id}: {str(e)}")
```

Comment:
As John already commented, there's no need to check all the daemons, you can use `mon` for example. Most of the time all of them will have the same version. I find the function a little bit complicated and hard to read, I'll suggest dividing it into two functions and have the main one call the other:

```
def get_python_version():
def retrieve_python_version_from_container(container_id):
```

Reply:
I have broken down the function into two parts now. Also it retrieves the python version from any one of the containers and returns instead of looping through all of them.